Rework infoLOS shader. Use a maximum intel based approach instead of additive.

DevJac opened the pull request:
# Overview

This PR simplifies the `infoLOS` shader and renders the fog of war based upon the maximum level of intel instead of the purely addition based approach of the old shader (I'll explain why this additive approach was a problem below.)

Also, the old shader was doing some really weird things. I believe the first evidence of this is how much simpler this shader is than the old one while still achieving very similar results. Another small example is the line `mix(diagLines2(texCoord), diagLines1(texCoord), heightStatus)` from the old shader, in which `heightStatus` could be -3 (negative three); that's quite unusual. The old shader also had unreachable code which I removed.

My goal in doing this was to make the areas that are not covered by radar more visible. To achieve this, I increased the width of the so-called "terra-incognita" lines. I also (ab)used the radarColor1 to allow the darkness of the terra-incognita lines to be adjusted outside of the shader. I will personally be adjusting this in a widget, but tried to leave the default colors and shader subtle like the existing one.

# Why a maximum intel approach is better than an additive approach.

A problem with the old shader is that every form of intel adds to the brightness. You want radar to add some brightness to the map to indicate coverage, and, of course, you also want direct line-of-sight to add brightness. The problem then is that radar coverage makes the areas with direct LOS _even brighter_. This is goofy.

To illustrate the problem, imaging you have several grunts spread loosely around the screen. They have direct LOS over their area and so that area is fairly bright; there is no fog of war around your grunts. Now imagine someone builds a T3 radar half-way across the map, and now the area around your grunts becomes even _brighter_. That's weird. A radar half-way across the map adds no intel to the area with your grunts, why should it become brighter?

This is how the additive approach must be. If the radar contributes significant brightness, and direct LOS contributes significant brightness, then together they are too bright.

This shader uses maximum instead of addition. So someone building a T3 radar across the map will brighten the appropriate dark areas, but it will _not_ brighten the areas where you already have direct LOS. This way both radar and direct LOS can contribute a significant amount of brightness without adding to become too bright.

# Screenshots

Here's a screenshot comparing the old and new shader. The first screenshot shows the old shader.

It's worth noting that the first screenshot demonstrates a bug with the old shader. If you look closely at the water covered by radar, you can see that there are still subtle terra-incognita lines that were not fully removed. I think this is because the old shader was returning values greater than 1.0 for `gl_FragColor`. This bug does not occur with the new proposed shader.

![screen00003](https://github.com/beyond-all-reason/Beyond-All-Reason/assets/5282489/e61915ed-2385-40e8-8f4a-f6abb2a9f75a)
![screen00007](https://github.com/beyond-all-reason/Beyond-All-Reason/assets/5282489/e308bead-6025-4a3b-a9ad-352332c36299)

DevJac committed:
Simplify LOS; give full LOS to both air and ground units

DevJac committed:
Remove radarAnim, it's not used

DevJac committed:
Remove hex2, it's not used

DevJac committed:
Simplify diagLines

DevJac committed:
Rewrite infoLOS shader; use maximum instead of addition

DevJac committed:
Adjust colors for new shader

DevJac commented:
This probably fixes https://github.com/beyond-all-reason/Beyond-All-Reason/issues/1817

DevJac committed:
Adjust colors to match previous look and feel; draw air LOS the old way

Beherith commented:
Have you perhaps seen my reimagings in https://github.com/beyond-all-reason/Beyond-All-Reason/commit/df06d635263b4653b63e80365820226dfc3a7c19 ?

Beherith commented:
@icexuick have a looksie here

DevJac committed:
Stop (ab)using radarColor1 for the diagonal line color

DevJac commented:
I see that the old shader code is duplicated in https://github.com/beyond-all-reason/Beyond-All-Reason/commit/df06d635263b4653b63e80365820226dfc3a7c19 ; I don't understand why.

DevJac committed:
Fix whitespace

sprunk commented:
One of your major premises is false, radar is not obsoleted by sight because it can see cloak where sight cannot.

DevJac commented:
@sprunk technically true. As you correctly pointed out, LOS _without_ radar coverage and LOS _with_ radar coverage are different in some cases, but notably, the game currently doesn't distinguish the two visually; it's never been a problem before, so I hope it wont be considered a problem now. Also, there is a limited amount of contrast available, and using some of that contrast on the edge case of cloaked units isn't worth it IMO.

DevJac commented:
Here's some additional screenshots of the difference. These are heavily compressed, so don't judge too closely, just look at the overall effect. Also, the difference pictured here is more extreme than what I've done in this PR.

![Screenshot from 2024-02-15 13-33-22](https://github.com/beyond-all-reason/Beyond-All-Reason/assets/5282489/d500aaf5-c8e9-4f76-a8f1-6af67377d16f)
![Screenshot from 2024-02-15 13-34-09](https://github.com/beyond-all-reason/Beyond-All-Reason/assets/5282489/28d94800-1517-4108-ac6f-d5633790b420)

icexuick commented:
My only 2 'things' to talk about are:

1. **Cloaked units**: With the old shader you could typically see this, though the effect is very subtle (radar+LOS vs LOS only) and i'd be surprised how many could tell the difference at a glance.
2. **Subtle immersion** - where venturing out into new areas, without proper radar coverage, will have a (subtle) effect because it's less bright compared to fully radar-covered.

This PR does do what it says, it simplifies the LOS layers and also cleans up some strange/goofy effects.
I think in general this PR is better. 

+ Better/more consistent contrast between in-view or no-view
+ Removes noise/overlap between radar | radar&LOS | LOS only
+ Should improve consistency behavior/looks on both bright & dark maps

Things I'd like to see improved:

- Diagonal lines maybe too big on zoom in - perhaps look/try a system with 'fixed' diagonal lines (regardless of zoom)
- Diagonal lines seem to show inside of radar range (should be a clear 'line' that aligns perfectly with the radar-range circle.

![image](https://github.com/beyond-all-reason/Beyond-All-Reason/assets/23189106/86e47004-3a9c-4f3a-b9ef-4a7fb160ad94)

- Ideally I'd like to add sliders to settings so people can tweak the presence/darkness of both radar and no-radar (diag lines) visibility
- Current setup in this PR is too dark for the map Forge

WatchTheFort commented:
>     * Ideally I'd like to add sliders to settings so people can tweak the presence/darkness of both radar and no-radar (diag lines) visibility

This is something we need to take care of ourselves, it is not the player's responsibility to adjust the shaders for each map. If we need to create a per-map config, then that's what we'll need to do.

I have never been a fan of the diagonal lines, but that is outside the scope of this PR.

DevJac commented:
Here's some screenshots from Forge. The brightness is not that different IMO.

### With this PR:

![Screenshot from 2024-02-25 15-56-07](https://github.com/beyond-all-reason/Beyond-All-Reason/assets/5282489/124b6db9-742d-4b36-aafa-90f974de70b3)
![Screenshot from 2024-02-25 15-56-27](https://github.com/beyond-all-reason/Beyond-All-Reason/assets/5282489/5ab8e0ee-639a-4351-884e-ec106cf91027)

### Without this PR (this is the default game, with 100% fog of war opacity):

![Screenshot from 2024-02-25 15-59-07](https://github.com/beyond-all-reason/Beyond-All-Reason/assets/5282489/ceae0c04-085c-4b0c-ae22-49ddfc7661c6)
![Screenshot from 2024-02-25 16-00-17](https://github.com/beyond-all-reason/Beyond-All-Reason/assets/5282489/dccfd7fe-e6ec-40ab-ad72-35f01f594f1f)

icexuick commented:
Not sure why my Forge game shots are much darker.
Are you on Nvidia or AMD?

Regarding your other feedback @DevJac:

- Regarding the sharp edges: I don't need exact sharp edges, I also know of the lower-res maps. But what triggers me is that it's such far off. Maybe you can dig in and find something to improve this a bit. Ivand might be able to help here.
- Other remarks are good/fair - I don't see any major deal breakers - though I want to tweak it a bit and test/check bright/dark maps more.

DevJac commented:
Scratch that, I don't know what I was thinking. I'm using an Nvidia GPU on Linux.

icexuick commented:
Was jamming-range also like this in original?

![image](https://github.com/beyond-all-reason/Beyond-All-Reason/assets/23189106/1ff75914-fdd6-4cd9-ae03-e3f8906f5e59)

icexuick commented:
First test on this map I think that difference between LOS and Radar is (too) small.

![image](https://github.com/beyond-all-reason/Beyond-All-Reason/assets/23189106/494f6026-1a0a-4788-8771-5a659f88230f)

icexuick commented:
I can confirm that the radar on the dotted line is indeed too large, and the LOS shader seems to be on point.

WatchTheFort commented:
I think the big problem we have with differentiating areas from one another is that we use a gradient where they meet, instead of a hard border. That problem isn't specific to this pull request, but I think solving that will go a long way to solving this issue, no matter how we style the different areas.

DevJac commented:
@WatchTheFort: I believe the radar coverage (and jammer, etc) is stored in a texture that is lower resolution than the map. That is why the boarder is fuzzy, because we don't have the information needed to draw a perfectly sharp boarder. There would be a performance cost if we store that data in a higher resolution texture.

DevJac commented:
@icexuick What map is that? I want to compare the radar / los contrast with the base game. As before, I suspect we will find that the base game already has a difficult to see transition between radar / los, but it will be good to check.

Actually, did we ever resolve this? I showed screenshots (from my computer) that showed little difference between this shader and the default game, but it sounds like you were seeing something else? Are we seeing the same things?

I'm open to talking on Discord and sharing screens or whatever.

icexuick commented:
> @WatchTheFort: I believe the radar coverage (and jammer, etc) is stored in a texture that is lower resolution than the map. That is why the boarder is fuzzy, because we don't have the information needed to draw a perfectly sharp boarder. There would be a performance cost if we store that data in a higher resolution texture.

Well i think hardlines will not look that 'pretty'.
Example:
![image](https://github.com/beyond-all-reason/Beyond-All-Reason/assets/23189106/39cb4da9-55bc-4676-b3d3-dae57a378bc0)

DevJac commented:
I don't think the hard lines look that bad, but I don't know how to change them. Maybe half way between the current smoothing and what we currently have?

icexuick commented:
Yes, I might be wise to jump in a VC / screenshare to discuss some of the topics.
Indeed it seemed pretty similar, though I was checking the (older) screenshots in this thread, and seeing similar results in my game.

Is there an option to 'sharpen' the edge with your new shader?

icexuick commented:
PS. please make a thread on Discord so we can channel all feedback there.

DevJac commented:
Pinged you on Discord.

But I want to reply to your feedback about that map and the low los / radar contrast.

I'm calling the 3 intel levels "los" / "radar" / "fow". "fow" meaning no radar coverage. (There are also things like los with radar and los without radar, but we've talked about that previously in this PR. There's only so many information channels available.)

That's an unavoidable trade-off here. We want los to remain roughly the same brightness, and fow to remain roughly the same brightness, and radar to be somewhere in between.

Currently radar has the same brightness as fow. That's my complaint and the reason I made this shader, there is no contrast between radar and fow areas.

To fix this I moved the radar brightness to be more in the middle between the two extremes. The brightness follows a roughly linear progression from los -> radar -> fow.

But, moving the radar brightness adds to one contrast while taking away from the other. In this case, the contrast between los and radar is reduced.

Let me illustrate it with some ascii art:

```
From most to least bright:

Current situation:  | los ------------------------------------------------ radar - fow |
In this shader:        | los ----------------------- radar -------------------------- fow |
```

Yes, the contrast (color distance) between los and radar has decreased, but the contrast between radar and fow increased.

I think I've beat this point to death now.

Of the two, I prefer having the radar contrast closer to los, because los is usually quite predictable, and there are also los lines that can be turned on in the settings (I have them on).

WatchTheFort commented:
The fact that there are LOS rings at all is completely bonkers. There is zero reason they should ever exist.

sprunk commented:
Units can be in a position where their LoS range is not evident:
* a unit can be in your base where everything is covered by LoS anyway so you don't see what the unit is contributing.
* a unit can be in a ditch where terrain blocks its LoS from all sides so you can't see how far it could reach if you get out.
* a unit could be under a status effect (your modding support is currently fake but you claim eventually it won't be) that reduces LoS and you might want to know its real LoS.
* you can learn still learn the value from a number in a stats spreadsheet in each case but that's lame.

icexuick commented:
> @icexuick What is the status of this? Is more work on it needed? Are we planning to merge it to the game?

Thats a good question - this has slipped through my fingers.

icexuick commented:
I tried running this, but I think it's not working anymore.
I would have to ask @DevJac to check and/or maybe update with latest game-code I presume?

WatchTheFort commented:
Can we please get rid of `L` and `;` functionality? It's our job to ensure the game is visible no matter which map and infoLOS state, not the players.

DevJac commented:
It's been a long time since I played BAR, and has been over a year since I looked at the code in this PR. I'm not interesting in working on this or contributing to BAR at the moment. If you want to close this PR, I understand.

icexuick commented:
> Can we please get rid of `L` and `;` functionality? It's our job to ensure the game is visible no matter which map and infoLOS state, not the players.

Well it's harder than you think - as it's a really sensitive system strongly depending on map diffuse/color/brightness.

`L` functionalities should remain (for people that don't want to play with this)
`;` functionality is the switch between 2 modes (with and without radar+los/radar-los)

I think in general most casual players, enjoying the game and immersion will want full LoS effect.
Pro players probably want to see the map optimally, and as little darkening as possible.
We can't unite these, so we need to toggle option(s).

I pretty swamped at the moment to look into this PR - though I like what @DevJac has done a lot!
I anyone else wants to pick this up, feel free to.

WatchTheFort commented:
@icexuick Has there been any progress made on this?

icexuick commented:
No, not that I know. 
I like the explanation and idea.

I do like the idea that exploring far outside your base with some units, having a slightly dimmer LOS because it's outside of your 'safe' zone (covered with radar) I do however see that this PR does clean things up.

Having a max brightness seems logical. @Beherith @PtaQQ @Ruwetuin what do you think?

DevJac committed:
Fix opacity setting, it affects fog of war darkness again

DevJac committed:
Move more math into diagLines function in shader

DevJac committed:
Keep the default 0.88 fow opacity, no reason to change it

DevJac commented:
Forge comparisons:

Current game, with 100, 50, and 0 fow opacity:

<img width="1263" height="847" alt="current_game_100" src="https://github.com/user-attachments/assets/b3f00e24-cac8-4a86-9b7f-1ad20054acb1" />
<img width="1263" height="847" alt="current_game_50" src="https://github.com/user-attachments/assets/04060923-6e42-485a-b6a7-9c8e0215b84c" />
<img width="1263" height="847" alt="current_game_0" src="https://github.com/user-attachments/assets/622de5b1-7920-4a34-bcd0-20e0b5ed748a" />

After the changes in this PR, it looks like this, again with 100, 50, and 0 fow opacity:

<img width="1263" height="847" alt="after_change_100" src="https://github.com/user-attachments/assets/2f366c28-3cd6-4b31-8b26-7d9a2b580cb4" />
<img width="1263" height="847" alt="after_change_50" src="https://github.com/user-attachments/assets/bb558b7c-8e95-4637-b198-73c83564b2c1" />
<img width="1263" height="847" alt="after_change_0" src="https://github.com/user-attachments/assets/a2d97d37-1c48-4739-876e-afbab35aed47" />

Keep in mind that those extremely dark areas only happen (in both the current game and PR) if the players choose to lower their fow opacity. It's a choice players get to make.

TheChronographer commented:
Is it meant to leave lingering los brightness? What could this light spot be indicating? 
<img width="1382" height="942" alt="image" src="https://github.com/user-attachments/assets/52b0f5b7-8f4e-4f19-ab60-b1da4735b015" />

You can clearly see a path the grunt walked along, 'newly discovered land'? But that has no gameplay effect. 

Also comparing master to this PR:
<img width="1114" height="1166" alt="image" src="https://github.com/user-attachments/assets/2574e983-6eb1-49ee-a249-4c844faefb9c" />

I find the radar no-los brightness is closer to no-radar no-los brightness. Which is the opposite of some of the above comment. But maybe I didn't understand the changes over the years. Yours looks way better on forged though, so likely an effect of me having 0.88 fog.

DevJac committed:
Removed most code and options from gfx_los_colors

A lot of this code was unneeded.

For example, I removed the specDetected variable and related
functionality and everything still works in the game.

I also removed the radar vizualization customizations, because most
players don't know about these, and the keys are not bound in
grid. Only the "legacy" bindings include some (but not all) of the
options.

This will be a good simplified base we can build up new functionality
on top of if desired.

WatchTheFort commented:
@icexuick

icexuick commented:
I'll try to squeeze in time next week for this.
I'm sorry this has taken so long!

icexuick committed:
Update to infoLOS

Fixed radar range (GUI) dotted range to be more correct to actual radar range
Added new GL4 diag-lines for no-radar coverage - based on LOS View GL4 interpolation/mask
Removed old/blurry infoLOS.lua diagonal lines
Added slider to control opacity of the new GL4 Diagonal lines.

Please also test on AMD

icexuick commented:
Okay, had to pick this one up.
I tried to fix / align the gui/green radar with the diagonal lines, but that was a dead-end street.

I updated the radar range to be a bit more accurate.
And wrote a new diagonal lines GL4 system to fix/replace the old method, which was too blurry (low-res) both in the lines as in the mask.

Bonus:

- Better / more smooth visually more accurate diagonal lines
- Better aligned with the actual GUI radar range rings
- Sharper lines/texture
- You can now define the strength of the diagonal lines real-time in settings

Please test this on AMD GPU as well - though I guess this should work.
I like the other settings.

icexuick commented:
@DevJac ^^

icexuick commented:
Here a short preview of the stuff I added.

https://www.youtube.com/watch?v=XVjSbk23MHg

icexuick commented:
one thing I could not replicate was on the map Forge.
Mine is much much darker.
Maybe Nvidia/AMD problem?

icexuick commented:
@Ruwetuin can you check/test this one and see if you have any issues/problems?

Ruwetuin commented:
when I start the game the stripes also displayed within radar range, until I toggled los (L key), after that it was shown as in your youtube video.  until you restart the game and its back again :/

icexuick commented:
> when I start the game the stripes also displayed within radar range, until I toggled los (L key), after that it was shown as in your youtube video. until you restart the game and its back again :/

I can't reproduce.
Could it be that you have some custom shaders running in your install?

do you have a file in your install here? `data/shaders/GLSL/infoLOS.lua`

Ruwetuin commented:
my bad it didnt load the PR properbly before. it works now, although i dont entirely like how big/clear the striping is now

WatchTheFort commented:
Is this at the point where the only work left to do is tweaking configured values to get the aesthetic we want?

icexuick commented:
I think so yes.
The striping can be fully adjusted in terms of size/width, distance between lines, speed, opacity even the blurryness.
I'd prefer to align on a fixed width/distance - and we can make the opacity a settings/slider in settings.

Ruwetuin committed:
Merge branch 'master' into master
